### ISSUE TITLE
OSX 10.11.3 - XCode (v7.2.1) Build Fails, 32/64-bit issue

### ISSUE THREAD

**Issue by rsnewcomb** [philburk/hmsl#73]
XCode build fails with 32 warnings, 1 error (target defaulted to 10.4; when changed to 10.11, same issues resulted).  I am attaching a screenshot of the directory structure I am using for the build, just in case you see a problem at that level that would cause the build to fail
![screen shot 2016-02-28 at 3 09 18 pm](https://cloud.githubusercontent.com/assets/17369962/13381601/c3b742ea-de2d-11e5-88b1-d5ff0669cc31.png)
:

hmsl Group
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_inner.c
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:322:9: Implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int'
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:335:26: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:336:26: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:405:20: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:407:39: Implicit conversion loses integer precision: 'unsigned long' to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:408:41: Implicit conversion loses integer precision: 'unsigned long' to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:409:24: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:433:22: Implicit conversion loses integer precision: 'ExecToken' (aka 'unsigned long') to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:451:22: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:452:23: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:464:20: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:471:19: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:489:31: Implicit conversion loses integer precision: 'long long' to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:520:9: Implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int32_t' (aka 'int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:583:11: Implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:692:11: Implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_save.c:709:11: Implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_gui.c
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_gui.c:63:17: Cast to 'char *' from smaller integer type 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_gui.c:64:53: Cast to 'char *' from smaller integer type 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_gui.c:77:29: Cast to 'char *' from smaller integer type 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_gui.c:153:4: Cast to 'int32_t *' (aka 'int *') from smaller integer type 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_gui.c:154:4: Cast to 'int32_t *' (aka 'int *') from smaller integer type 'uint32_t' (aka 'unsigned int')
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_midi.m
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_midi.m:46:35: 'AbsoluteToNanoseconds' is deprecated: first deprecated in OS X 10.8
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/hmsl_midi.m:186:29: 'NanosecondsToAbsolute' is deprecated: first deprecated in OS X 10.8
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_guts.h
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_guts.h:571:9: 'MIN' macro redefined
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/pforth-xcode/csrc/pf_guts.h:572:9: 'MAX' macro redefined
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/HMSLWindowDelegate.m
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/HMSLWindowDelegate.m:16:15: Unused variable 'window'
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/HMSLRunner.m
/Users/rsn/Documents/HD_DATA/PROJECTS/MITA2/PHMSL/hmsl-master/native/osx/src/HMSLRunner.m:77:12: Implicit conversion loses integer precision: 'cell_t' (aka 'long') to 'int'
Mv: rename pforth.dic to /Users/rsn/Library/Developer/Xcode/DerivedData/HMSL-OSX-dmbirwwokmaietggeruaffsdmujc/Build/Products/Debug/pforth.dic: No such file or directory

**Comment by andrewcsmith** [comment]
Yep, the error is that it can't find pforth.dic. Probable cause is that the forth dictionary is not compiling. This is one of the steps in the hmsl.app bundle.

In xcode, go to the **Report navigator** and check out the build report for the hmsl (app bundle) target. Let me know what that says—the error should be there.

You should be able to work around it by deleting the final "Run Script" block from the Build Phases section of the settings. But I bet that it just can't find one of the previous pforth dictionaries.

**Comment by rsnewcomb** [comment]
Here is what Report Navigator reveals:

![screen shot 2016-02-29 at 7 06 27 pm](https://cloud.githubusercontent.com/assets/17369962/13413340/c67f90aa-df17-11e5-9185-137c8bbb7640.png)

**Comment by andrewcsmith** [comment]
key point there is `error in pfLoadDictionary - cell size mismatch between .dic file and code` — @philburk  any ideas?

**Comment by philburk** [comment]
The report says "Cell Size = 4". That message is about the dictionary. So the dictionary pforth_system.dic is for a 32-bit Forth.
But up above it says "PForth V27-LE/64". That means the pForth kernel is 64-bit!
So they are incompatible. The pForth kernel should be 32-bit.

Two questions:
1) Where is that 64-bit pForth kernel coming from?
2) How did you generate a 32-bit pforth_system.dic from a 64-bit pForth? There must also be a 32-bit pForth somewhere.

**Comment by rsnewcomb** [comment]
Just want to quadruple check that there is nothing wrong with how I have pForth and HMSL downloaded and setup on my laptop.  Please see the screenshot at the top of this Issue.  Phil, you had said the installation instructions on GitHub were out of date and gave me these steps to follow, which I believe I have:

On El Capitan, download pforth-master.zip and hmsl-master.zip and put them in the same folder.
unzip them both by double clicking

rename "pforth-master" to "pforth-xcode"
Open XCode project in hmsl-master/native/osx
Select the second "hmsl" target.
Build.

Click on the Folder icon
Right click on "Product / hmsl.app" and select "Show in Finder"
Drag the two files "hmsl.app" and "pforth.dic" into "hmsl-master/hmsl/" folder.

Double click on hmsl.app
Enter 'y' to initialize
Enter: SHEP

You should see the Shape Editor

In Logic or Simple Synth, select the HMSL MIDI output

**Comment by philburk** [comment]
I believe that is the new procedure. It worked for me on El Capitan.
I assume you're failing at the "Build." step.

I notice from your report that the Build failed on 2/28 at 3:10 PM. But pForth was built on 2/27 at 11:32 PM. So it is picking up old file. Try doing a "Product / Clean" and do a fresh build.

For some reason you are running a 64-bit pForth. Why?
Please check the Build Settings by clicking on the project. Are both HMSL-OSX and hmsl targets set to 32bit?

<img width="1133" alt="hmsl_32bit_build" src="https://cloud.githubusercontent.com/assets/5175913/13432673/857bb414-df83-11e5-8036-9ea69526d4b4.png">

**Comment by andrewcsmith** [comment]
Okay, thanks @philburk. The 64-bit pForth will be the one compiled as HMSL (that's the version used in the build script). Looking closer at your warnings, it looks like you have XCode set up to compile as 64-bit. XCode *usually* prompts you to make that change, but since HMSL doesn't work with 64-bit architectures you want to ignore that prompt. Perhaps you clicked "ok" instead of "cancel" at some point?

I can confirm that a fresh download of the zip compiles just fine on my machine, and works with the packaged `pforth_system.dic`.

**Comment by rsnewcomb** [comment]
Cleaned the Project.  Build.  Got this set of three errors:

![screen shot 2016-03-01 at 9 27 18 pm](https://cloud.githubusercontent.com/assets/17369962/13448891/84319a4e-dff4-11e5-85b9-2e9ea4866750.png)

**Comment by rsnewcomb** [comment]
Progress!  I was able to download both zips, examine the XCode project settings - THIS time I unchecked all Build defaults - and the Build was fine:

Upon running the .app file, if I initialize with no memory allocation changes, I am fine.  However, I always will be allocating more memory for both code and headers.  On the 'save-forth' command I received this error when trying to to so:

![screen shot 2016-03-17 at 8 03 58 pm](https://cloud.githubusercontent.com/assets/17369962/13864895/e41c7472-ec7b-11e5-9b67-48f56f2abe89.png)

The issue is that using the binary received from Phil on 2/20/16, I have been able to do this with no problem, and continued to run legacy H4th code through the interpreter to build the open issues list.  Suggestions on what might be the difference between the locally and remotely built apps?

Robert
